### DISCUSSION TITLE
Second argument to jsPsych.randomization.shuffleNoRepeats() must be a function

### DISCUSSION THREAD

**Discussion by mcvc968** [jspsych/jsPsych#1822] [Q&A]
Hi,

I am trying to loop through an array ('base') 3 times and generate 3 randomised arrays with no repeating neighbours. I am trying to do this using jsPsych.randomization.shuffleNoRepeats . I keep getting the error 'Second argument to jsPsych.randomization.shuffleNoRepeats() must be a function' and I was wondering if I am using it wrong syntax?  I am attaching the relevant portion of the code below. Thank you for your help in advance. 

```js
var base = [
      /*A*/
      {stimulus:"emotion_sounds/R4.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", man], choices: ['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 'a' }, },
      /*S*/
      {stimulus:"emotion_sounds/R3.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", lamp], choices: ['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 's' }, },
      /*L*/
      {stimulus:"emotion_sounds/R1.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", woman], choices: ['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 'l' }, },
      /*K*/
      {stimulus:"emotion_sounds/R2.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", chair], choices:['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 'k' }, },
      /*A*/
      {stimulus:"emotion_sounds/R4.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", man], choices: ['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 'a' }, },
      /*S*/
      {stimulus:"emotion_sounds/R3.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", lamp], choices: ['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 's' }, },
      /*L*/
      {stimulus:"emotion_sounds/R1.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", woman], choices: ['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 'l' }, },
      /*K*/
      {stimulus:"emotion_sounds/R2.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", chair], choices:['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 'k' }, },
      /*A*/
      {stimulus:"emotion_sounds/R4.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", man], choices: ['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 'a' }, },
      /*S*/
      {stimulus:"emotion_sounds/R3.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", lamp], choices: ['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 's' }, },
      /*L*/
      {stimulus:"emotion_sounds/R1.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", woman], choices: ['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 'l' }, },
      /*K*/
      {stimulus:"emotion_sounds/R2.wav", prompt: ["<h1 style= background-color:powderblue;>Random</h1>", chair], choices:['a','s','k','l'], data: { test_part: 'Rand_seq', correct_response: 'k' }, },
      //{stimulus: fixation},
  ];

var randomised = [];


  for (var i = 0; i < 3; i++) {

      var repeatedSet = jsPsych.randomization.repeat(base, 1);
      var repeatedSet_flat = repeatedSet.flat();
      var randomise = jsPsych.randomization.shuffleNoRepeats(repeatedSet_flat, function (a, b) { return a.stimulus === b.stimulus }); //, 12)
      randomised.push(randomise)
      random_flat = randomised.flat()
};
```

**Comment by jodeleeuw** [comment]
Hi @mcvc968,

I can't replicate the issue. I used your code as you pasted it. Here's the test site: https://bristle-spangled-syzygy.glitch.me/.

If you open the console you can see the value of `randomised`, which looks correct to me.

What version of jsPsych are you using?

**Comment by mcvc968** [comment]
Hi, 

Thank you for your response. I'm sorry, you're right! The problem is actually further down the script. In it I am trying to shuffle  'all_tests_shuffled_2' such that there are no more than three repeated neighbours and that is where I get the error about the second argument being a function. 

Sorry for the double question!

var myarray_2 = [tests_A_flat, tests_B_flat, tests_rand_flat];

 var all_tests_shuffled_2 = jsPsych.randomization.repeat(myarray_2, [9,9,2]);
  var shuffled_noreps_2 = jsPsych.randomization.shuffleNoRepeats(all_tests_shuffled_2, function (a,b,c) { return a.task === b.task  === c.task });

    for (var curr_test = 0; curr_test < shuffled_noreps_2.length; curr_test ++)
        {
            // Take one object of length 2 and convert it to an array of length 2
            shuffled_noreps_2[curr_test] = Object.values(shuffled_noreps_2[curr_test]);
        }
    // Now that we have an array of arrays instead of an an array of objects, we can flatten it
    all_tests_shuffled_flat_2 = shuffled_noreps_2.flat();
